File: src/domain/inspection/value_objects/InspectionMasterCatalog.js
/* exported InspectionMasterCatalog */

class InspectionMasterCatalog {
  /**
   * @param {Map<string, InspectionMasterItem>} itemsByAsin
   */
  constructor(itemsByAsin) {
    this._itemsByAsin = itemsByAsin || new Map();
  }

  /**
   * @param {string[]} asins
   * @returns {InspectionMasterCatalog}
   */
  filterByAsins(asins) {
    const set = new Set((asins || []).map(a => String(a || '').trim()).filter(Boolean));
    const next = new Map();
    for (const [asin, item] of this._itemsByAsin.entries()) {
      if (set.has(asin)) {
        next.set(asin, item);
      }
    }
    return new InspectionMasterCatalog(next);
  }

  /**
   * @param {string} asin
   * @returns {boolean}
   */
  has(asin) {
    return this._itemsByAsin.has(String(asin || '').trim());
  }

  /**
   * @param {string} asin
   * @returns {InspectionMasterItem|null}
   */
  get(asin) {
    return this._itemsByAsin.get(String(asin || '').trim()) || null;
  }

  /**
   * @returns {number}
   */
  size() {
    return this._itemsByAsin.size;
  }

  /**
   * @returns {string[]}
   */
  asins() {
    return Array.from(this._itemsByAsin.keys());
  }
}

File: python/src/domain/inspection/value_objects/inspection_master_catalog.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from domain.inspection.entities.inspection_master_item import InspectionMasterItem


class InspectionMasterCatalog:
    def __init__(self, items_by_asin: dict[str, InspectionMasterItem]) -> None:
        self._items_by_asin = dict(items_by_asin)

    def filter_by_asins(self, asins: list[str]) -> InspectionMasterCatalog:
        asin_set = set(asins)
        filtered = {k: v for k, v in self._items_by_asin.items() if k in asin_set}
        return InspectionMasterCatalog(items_by_asin=filtered)

    def has(self, asin: str) -> bool:
        return asin in self._items_by_asin

    def get(self, asin: str) -> InspectionMasterItem | None:
        return self._items_by_asin.get(asin)

    def size(self) -> int:
        return len(self._items_by_asin)

    def asins(self) -> list[str]:
        return list(self._items_by_asin.keys())
